pmldr/pmldr/fs.asm

```diff
@@ -0,0 +1,333 @@
+[BITS 16]
+
+LoadDiskSectors:
+	push ebp
+	mov ebp, esp
+;Loads disk sectors into memory, at (ds):0000. Overwrites this memory and clobbers registers.
+;Arguments: bx = number of sectors to read in, cx = (16-bit) sector offset. Assume 1 sector = 512 bytes.
+;set up a disk address packet structure at 0x300:000. stos* stores at es:di and increments. int13 needs struct in ds:si
+	;this is a new version that loads one sector at a time, to accomodate bios quirks (see https://stackoverflow.com/questions/58564895/problem-with-bios-int-13h-read-sectors-from-drive)
+	cld		;make sure we are writing forwards
+	push es
+	push ds
+	mov ax, 0x300
+	mov es, ax
+
+	mov di, 0x00
+	mov al, 0x10	;16 byte packet
+	stosb
+	xor al,al	;field always 0
+	stosb
+	mov ax, 1
+	stosw
+	xor ax,ax	;write to sector origin
+	stosw
+	mov ax, ds	;segment to write
+	stosw
+	mov ax, cx	;we want to read one sector from offset of sector 4, but this is little-endian. Write 48 bits in 4 words - LSW first
+	stosw
+	xor ax,ax
+	stosw
+	stosw
+	stosw		;now we are ready
+	mov ax, es
+	mov ds, ax
+	xor ax, ax
+
+	_load_next_sector:
+	mov si, 0x00
+	mov ah, 0x42
+	mov dl, 0x80					;FIXME: hardcoded disk number!
+	int 0x13
+	jc fat_load_err					;this hangs the bootloader so it does not really matter that we leave two values on the stack
+	dec bx							;track how many sectors we want to load in bx
+	test bx,bx
+	jz _load_disk_sectors_done
+	add word [es:0x06], 0x020		;each block is 512 bytes = 0x200. Increment the destination ptr by this much. FIXME this block size should be derived from the GET DRIVE PARAMETERS (int0x15/0x48 command)
```

Comment:
This is a reasonable guess so can leave it for the time being; however it's unlikely to work when booting from CD